src/compiler/checker.ts

```diff
@@ -36458,9 +36534,50 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-        if (!(isTypeAny(rightType) || typeHasCallOrConstructSignatures(rightType) || isTypeSubtypeOf(rightType, globalFunctionType))) {
-            error(right, Diagnostics.The_right_hand_side_of_an_instanceof_expression_must_be_of_type_any_or_of_a_type_assignable_to_the_Function_interface_type);
+        if (!isTypeAny(rightType)) {
+            // if rightType is an object type with a custom `[Symbol.hasInstance]` method, then it is potentially
+            // valid on the right-hand side of the `instanceof` operator. This allows normal `object` types to
+            // participate in `instanceof`, as per Step 2 of https://tc39.es/ecma262/#sec-instanceofoperator.
+            const hasInstanceMethodType = getSymbolHasInstanceMethodOfObjectType(rightType);
+            if (hasInstanceMethodType) {
+                // avoid a complex check for every `instanceof` when the `[Symbol.hasInstance]` method has a single
+                // call signature that neither restricts nor narrows (via type predicate) the LHS value, e.g.
+                // `(value: unknown) => boolean`.
+                const cache = hasInstanceMethodType as HasInstanceMethodType;
+                if (cache.hasSimpleUnrestrictedSingleCallSignature === undefined) {
+                    const signature = getSingleCallSignature(hasInstanceMethodType);
+                    cache.hasSimpleUnrestrictedSingleCallSignature = !!signature && signature.parameters.length === 1 &&
+                        !!(getTypeOfSymbol(signature.parameters[0]).flags & TypeFlags.AnyOrUnknown) &&
+                        !!signature.resolvedReturnType &&
+                        !!(signature.resolvedReturnType.flags & TypeFlags.Boolean);
+                }
+                if (!cache.hasSimpleUnrestrictedSingleCallSignature) {
+                    // If rightType has a `[Symbol.hasInstance]` method that is not `(value: unknown) => boolean`, we
+                    // must check the expression as if it were a call to `right[Symbol.hasInstance](left)1. The call to
+                    // `getResolvedSignature`, below, will check that leftType is assignable to the type of the first
+                    // parameter.
+                    const syntheticCall = createSyntheticHasInstanceMethodCall(left, right, leftType, hasInstanceMethodType);
+                    const returnType = getReturnTypeOfSignature(getResolvedSignature(syntheticCall));
+
+                    // We also verify that the return type of the `[Symbol.hasInstance]` method is assignable to
+                    // `boolean`. According to the spec, the runtime will actually perform `ToBoolean` on the result,
+                    // but this is more type-safe.
+                    checkTypeAssignableTo(returnType, booleanType, right, Diagnostics.An_object_s_Symbol_hasInstance_method_must_return_a_boolean_value_for_it_to_be_used_on_the_right_hand_side_of_an_instanceof_expression);

```

Comment:
If we we check if the return type is assignable to `bool` here, why don't we check the return type in the implementations of `[Symbol.hasInstance]` methods?

Reply:
We have no way to validate a `[Symbol.hasInstance]` method on a given object without prompting, and its implementation only matters when it is going to be used as part of `instanceof`, just as we don't validate that a given `[Symbol.iterator]()` method is valid until it is used in places where a valid iterable is expected.